olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,244 @@
+    upload_to_cloud: bool = True
+    input_model_config: ModelConfig = None
+    hf_cache_path: str = os.environ.get("HF_HOME", None) or str(Path("~/.cache/huggingface").expanduser())
+
+
+class CloudModelInvalidStatus(str, Enum):
+    NO_MODEL_FILE = "NO_MODEL_FILE"
+    FAILED_CONFIG = "FAILED_CONFIG"
+
+
+class CloudCacheHelper:
+    def __init__(
+        self,
+        cache_dir: str,
+        account_url: str,
+        container_name: str,
+        hf_cache_path: str,
+        input_model_config: ModelConfig = None,
+    ):
+        try:
+            from azure.storage.blob import ContainerClient
+        except ImportError as exc:
+            raise ImportError(
+                "Please install azure-storage-blob and azure-identity to use the cloud model cache feature."
+            ) from exc
+        credential = get_credentials()
+        self.container_name = container_name
+        self.container_client = ContainerClient(
+            account_url=account_url, container_name=self.container_name, credential=credential
+        )
+
+        self.map_blob = "model_map.json"
+        self.local_map_path = Path(cache_dir) / "model_map.json"
+        self.local_map_path.parent.mkdir(parents=True, exist_ok=True)
+        self.cache_dir = cache_dir
+        self.output_model_path = Path(cache_dir) / "model"
+        self.cloud_cache_map = self.download_model_cache_map()
+        self.input_model_config = input_model_config
+        self.hf_cache_path = hf_cache_path
+
+    def update_model_config(self, cloud_model_path, model_config, input_model_hash):
+        logger.info("Updating model config with cloud model path: %s", cloud_model_path)
+        model_directory_prefix = f"{input_model_hash}/model"
+        blob_list = self.container_client.list_blobs(name_starts_with=model_directory_prefix)
+
+        for blob in blob_list:
+            blob_client = self.container_client.get_blob_client(blob)
+            local_file_path = self.output_model_path / blob.name[len(model_directory_prefix) + 1 :]
+            logger.info("Downloading %s to %s", blob.name, local_file_path)
+            local_file_path.parent.mkdir(parents=True, exist_ok=True)
+            with open(local_file_path, "wb") as download_file:
+                download_stream = blob_client.download_blob()
+                download_file.write(download_stream.readall())
+
+        model_config.config["model_path"] = str(self.output_model_path / cloud_model_path)
+
+        return model_config
+
+    def download_model_cache_map(self):
+        logger.info("Downloading model cache map from %s", self.map_blob)
+        blob_client = self.container_client.get_blob_client(self.map_blob)
+
+        if not blob_client.exists():
+            return {}
+
+        self.local_map_path.unlink(missing_ok=True)
+        self.local_map_path.touch()
+
+        with open(self.local_map_path, "wb") as download_file:
+            download_stream = blob_client.download_blob()
```

Comment:
Should `download_stream` use `with` to support auto close when done? It might probably close automatically when reaching EOF but being explicit might help.

Reply:
I was following the pattern from documentation here: https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blob-download-python#download-to-a-file-path

Reply:
Hmm, I guess `download_blob` doesn't actually return a stream - the returned object actually wraps multiple 'get data in range' calls. You could, however, replace the two calls with one [readinto](https://learn.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.storagestreamdownloader?view=azure-python#azure-storage-blob-storagestreamdownloader-readinto) call